### PR TITLE
clientv3/integration: Don't retry other endpoints when err == rpctypes.ErrAuthNotEnable 

### LINKED ISSUE #10430: Client return "context deadline exceeded" err with disable RoleAuth and the username and password are used
# etcd version
3.23
ETCD Cluster doesn't enable Role Auth, and the username and password are used in client.

# For example:
```ETCD_CONNECT_TIMEOUT = 3 * time.Second
var Etcd_dsn []string = ["192.168.1.10:2379","192.168.1.11:2379","192.168.1.12:2379"]
var Etcd_username string = "root"
var Etcd_passwd string = "pass"
cfg := etcdclient.Config{
		Endpoints: Etcd_dsn,
		TLS:       _tlsConfig,
		DialTimeout:ETCD_CONNECT_TIMEOUT,
		AutoSyncInterval: time.Second * 300,
		Username: Etcd_username,
		Password: Etcd_passwd,
	}
client, err := clientv3.New(cfg)
if err != nil {
	return err
}
```
# Two results:
1. It will work well with less endpoints
2. It will return `context deadline exceed` with many endpoints

# The rootcause of BUG:
1. When client code use username and password ,it will go to the code `err := c.getToken(ctx)` at client.go line 356.
2. In `getToken` function, it will not stop trying another endpoint until sucessfull if get error after call the `auth.authenticate`.
3. It will return `rpctypes.ErrAuthNotEnabled` after trying all endpoints if ETCD cluster don't use Auth.
4. When with less endpoints, the `ctx` will not get timeout after trying all endpoints.
5. When with many endpoints, the `ctx` will get timeout when after trying most endpoints. The error `rpctypes.ErrAuthNotEnabled` in the 311th line of client.go will turn to be `context deadline exceeded`
6. In the line client.go 365, it will be ignored when the err is not `rpctypes.ErrAuthNotEnabled` when handling the result of `c.getToken(ctx)`. It will not communicate normaly when get other error types. 

We can get a conclusion that it will not function as well as we wish when the number of endpoints is not sure. 

In my production , the ETCD cluster in the TEST environment has three endpoints which are deployed in different IDC , while it in PRODUCT environment has seven or nine endpoints which are deployed in different IDC. It will work well in the TEST but doesn't work in the PRODUCT when the code isn't changed.

The err `rpctypes.ErrAuthNotEnabled` can be ignored if the communication to ETCD works well just like in client.go 358 line . 
I think it can return directly when the function `getToken` get err `rpctypes.ErrAuthNotEnabled`. It don't need to try other endpoints.

### PR DESCRIPTION
**Pull request by cfc4n**
Fixes #10430

### CODE REVIEW

**Comment by xiang90** [comment]
@cfc4n can you add a test?

**Comment by cfc4n** [comment]
> @cfc4n can you add a test?

OK，please wait...

**Comment by cfc4n** [comment]
@xiang90 Is my test code correct?

File: clientv3/client.go [hunk 1]
```diff
@@ -310,6 +310,10 @@ func (c *Client) getToken(ctx context.Context) error {
 		var resp *AuthenticateResponse
 		resp, err = auth.authenticate(ctx, c.Username, c.Password)
 		if err != nil {
+			if err == rpctypes.ErrAuthNotEnabled {
+				// return err if err == rpctypes.ErrAuthNotEnable ,Don't retry other endpoints
```
Comment (spzala) [inline clientv3/client.go]: I would suggest to just say something like "// return err without retrying other endpoints" (to keep it clean as the if condition is self explanatory)

File: clientv3/example_auth_test.go [hunk 2]
```diff
@@ -111,3 +113,32 @@ func ExampleAuth() {
 	// Output: etcdserver: permission denied
 	// user u permission: key "foo", range end "zoo"
 }
+
+
+// fixes # https://github.com/etcd-io/etcd/issues/10430
```
Comment (spzala) [inline clientv3/example_auth_test.go]: Should remove this comment.

File: clientv3/example_auth_test.go [hunk 3]
```diff
@@ -111,3 +113,32 @@ func ExampleAuth() {
...
+func ExampleGetTokenWithoutAuth() {
+	var tmpEndpoints = []string{}
+	fmt.Println("it will append 100 times into {tmpEndpoints} with {endpoints}.")
```
Comment (spzala) [inline clientv3/example_auth_test.go]: Also, I don't think this print statement is needed.

File: clientv3/client.go (hunk 1 above)
Comment (cfc4n) [reply clientv3/client.go]: Sure,Thanks @spzala

Comment (spzala) [reply clientv3/client.go]: No problem @cfc4n and thank you for addressing the suggestions!

**Comment by mitake** [comment]
@xiang90 got it, I'll take a look probably this weekend.
@cfc4n the idea seems to be good, thanks! But could you fix the commit and coding style? The problems can be detected by executing `PASSES=fmt ./test` on your repository.

**Comment by cfc4n** [comment]
@mitake Done,I'll do it next time.thanks.

**Comment by xiang90** [comment]
@cfc4n 

Can you add the test as an integration test here: https://github.com/etcd-io/etcd/tree/master/clientv3/integration instead of an example? thanks!

**Comment by mitake** [comment]
@cfc4n could you fix errors from fmt test and build errors reported in https://travis-ci.com/etcd-io/etcd/jobs/173656226 ?

File: clientv3/integration/user_test.go [hunk 4]
```diff
@@ -110,3 +110,42 @@ func authSetupRoot(t *testing.T, auth clientv3.Auth) {
...
+	// "Username" and "Password" must be used
+	cfg := clientv3.Config{
+		Endpoints:   authapi.Endpoints(),
+		DialTimeout: 1 * time.Second, // make sure all connection time of connect all endpoint must be more DialTimeout
+		Username:    "root",
+		Password:    "123",
+	}
+
+	client, err = clientv3.New(cfg)
+
```
Comment (xiang90) [inline clientv3/integration/user_test.go]: defer client.Close()

File: clientv3/integration/user_test.go [hunk 5]
```diff
@@ -110,3 +110,42 @@ func authSetupRoot(t *testing.T, auth clientv3.Auth) {
...
+	switch err {
+	case nil:
+		t.Log("passes as expected, but may be connection time less than DialTimeout")
+	case context.DeadlineExceeded:
+		t.Fatalf("not expected result:%v with endpoint:%s", err, authapi.Endpoints())
```
Comment (xiang90) [inline clientv3/integration/user_test.go]: t.Errorf

File: clientv3/integration/user_test.go [hunk 6]
```diff
@@ -110,3 +110,42 @@ func authSetupRoot(t *testing.T, auth clientv3.Auth) {
...
+	case rpctypes.ErrAuthNotEnabled:
+		t.Logf("passes as expected:%v", err)
+	default:
+		t.Fatalf("other errors:%v", err)
```
Comment (xiang90) [inline clientv3/integration/user_test.go]: t.Errorf

**Comment by xiang90** [comment]
please squash commits into one with the right prefix "clientv3". thanks!

**Comment by xiang90** [comment]
@cfc4n 

Now you have 10 commits in one issue. You need to follow this:

https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

**Comment by cfc4n** [comment]
Got it ,thanks.

**Comment by xiang90** [comment]
@cfc4n 

hmm... there are now 12 commits in this PR.

Can you try

```
git reset --soft HEAD~12 # reset all commits but keep the changes
git commit -am "clientv3/integration: return err if err == rpctypes.ErrAuthNotEnable" # make a commit
git push -f xxx # force push to update the PR
```

File: clientv3/integration/user_test.go [hunk 7]
```diff
@@ -110,3 +110,42 @@ func authSetupRoot(t *testing.T, auth clientv3.Auth) {
...
+	case nil:
+		t.Log("passes as expected, but may be connection time less than DialTimeout")
+	case context.DeadlineExceeded:
+		t.Errorf("not expected result:%v with endpoint:%s", err, authapi.Endpoints())
+	case rpctypes.ErrAuthNotEnabled:
+		t.Logf("passes as expected:%v", err)
+	default:
+		t.Errorf("other errors:%v", err)
+	}
+	defer client.Close()
```
Comment (xiang90) [inline clientv3/integration/user_test.go]: well. if you still put client.Close() here, then no need to add defer. move this to line 138, right after the creation of the client instead.

**Comment by cfc4n** [comment]
@xiang90 Is it correct now?   But I'm also not understand the report in https://travis-ci.com/etcd-io/etcd/jobs/174849460 ,why it's failed  at [Job 666.4](https://travis-ci.com/etcd-io/etcd/jobs/174849460)\ [Job 666.9](https://travis-ci.com/etcd-io/etcd/jobs/174849465)\ [Job 666.11](https://travis-ci.com/etcd-io/etcd/jobs/174849467)  and, what can I do for this report?  It seems has nothing with my changes.

**Comment by xiang90** [comment]
the test failure has nothing to do with this RP. LGTM. Thanks for the contribution!

**Comment by jingyih** [comment]
cc @jpbetz

**Pull request merged**